core/species.cpp

```diff
@@ -8398,6 +8399,64 @@ void Species::__AddMutationsFromTreeSequenceToGenomes(std::unordered_map<slim_mu
 	free(variant);
 }
 
+void Species::__CheckNodePedigreeIDs(EidosInterpreter *p_interpreter)
+{
+	// Make sure our next pedigree ID is safe; right now it only accounts for pedigree IDs used by individuals, but maybe there
+	// could be nodes in the node table with genome pedigree IDs greater than those in use by individuals, in nonWF models.
+	// See https://github.com/MesserLab/SLiM/pull/420 for an example model that does this very easily.
+	
+	// Also, check for duplicate pedigree IDs, just in case.  __CreateSubpopulationsFromTabulation() does this for individual
+	// pedigree IDs; we do it for node pedigree IDs.  I decided to use a vector with std::sort() to check even though it is
+	// O(n log n), rather than a hash table for O(n), because I'm nervous about hitting a bad edge case with the hash table
+	// due to the nature of the values being inserted.  Shouldn't be a big deal in the grand scheme of things.
+	tsk_node_table_t &node_table = tables_.nodes;
+	tsk_size_t node_count = node_table.num_rows;
+	std::vector<slim_genomeid_t> genome_id_check;
+	
+	for (tsk_size_t j = 0; (size_t)j < node_count; j++)
+	{
+		tsk_size_t offset1 = node_table.metadata_offset[j];
+		tsk_size_t offset2 = node_table.metadata_offset[j + 1];
+		tsk_size_t length = (offset2 - offset1);
+		
+		if (length)		// allow nodes with no metadata, which might be carryover non-SLiM nodes
+		{
+			// but if a node has metadata, it must be SLiM metadata, I think; our schema guarantees that
+			if (length != sizeof(GenomeMetadataRec))
```

Comment:
I think this is inserting a new requirement on the loaded tree sequence that might well come back to bite us - previously, we didn't care what the metadata was for ancestral nodes.  I think you may be right about the metadata schema, but if we don't need to be more picky about metadata than we were before, let's not start?
```suggestion
		if (length != sizeof(GenomeMetadataRec))
```

Reply:
I realize there is now the possibility that there is metadata of the right length but the wrong structure, that errors, but this seems pretty unlikely.

Reply:
So, if the metadata length == sizeof(GenomeMetadataRec) then assume it is our metadata and go into it, if it is any other length then just ignore it?  I can do that, sure, although eww.

Reply:
But yeah, I see your point, and length is the only criterion we really have for checking whether metadata is ours or not, so there we are.